app/services/notification_subscriptions/toggle.rb

```diff
@@ -0,0 +1,94 @@
+module NotificationSubscriptions
+  class Toggle
+    attr_reader :current_user, :permitted_params
+
+    def self.call(...)
+      new(...).call
+    end
+
+    def initialize(current_user, permitted_params)
+      @current_user = current_user
+      @permitted_params = permitted_params
+    end
+
+    def call
+      toggle_subscription
+    end
+
+    def toggle_subscription
+      action = permitted_params[:action]
+
+      if action == "unsubscribe"
+        subscription_id = permitted_params[:subscription_id]
+        notification = NotificationSubscription&.find(subscription_id) unless subscription_id.nil?
+        destroy_notification(notification)
+      else
+        comment_id = permitted_params[:comment_id]
+        article_id = permitted_params[:article_id]
+        comment = Comment.find(comment_id) if comment_id.present?
+        article = Article.find(article_id) if article_id.present?
+
+        create_subscription(comment, article)
+      end
+    end
+
+    private
+
+    def destroy_notification(notification)
+      notification.destroy
+
+      if notification.destroyed?
+        { destroyed: true }
+      else
+        { errors: notification.errors_as_sentence }
+      end
+    end
+
+    def create_subscription(comment, article)
+      comment_article = comment && comment.ancestry.nil? ? comment.commentable : nil
+      notifiable = subscription_notifiable(comment, article, comment_article)
+      notifiable_type = if comment && article.nil? && comment_article.nil?
```

Comment:
Minor: this might be easier to read if the `comment`, `article` and `comment_article` were moved to memoized helper methods — `notifiable_type`, `subscription_config` and `notifiable` could also move that direction then, as well.